Revise AuthenticationSelectionResolver (#12102)

thomasdarimont opened the pull request:
Previously the order of the generated AuthenticationExecutionOptions
was partially based on the order of user credentials, and  not
on the order of the Authentication flow definition.

This change simplifies the logic of the AuthenticationSelectionResolver and
collects the Authentication options in the order they are defined in the
corresponding Authentication flow.

The new behavior is now easier to understand because the order of the
executions in the authentication flow is now followed exactly when
generating the authentication options.

Note: This might change the order in which authentication options
are presented for some users.

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

thomasdarimont commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/MultiFactorAuthenticationTest.java:
```diff
@@ -147,6 +147,12 @@ public void testAlternativeCredentials() {
             // like user with user-with-one-configured-otp. However OTP is preferred credential for him, so OTP mechanism will take preference
```
Acutally, I just saw this comment.

It seems that the order of credentials did indeed matter for determining the order of authentication options. 
However, AFAIK the user cannot control the order of credentials - only administrators can change the credential order.

I think relying on the order of stored credentials is fragile and very confusing. It would be better to rely on a deterministic order governed by the authentication flow.

mabartos commented:
@mposolda It'd be great if you'd have a chance to review it, or just share your thoughts. Thank you :))

thomasdarimont committed:
Revise AuthenticationSelectionResolver (#12102)

Previously the order of the generated AuthenticationExecutionOptions
was partially based on the order of user credentials, and  not
on the order of the Authentication flow definition.

This change simplifies the logic of the AuthenticationSelectionResolver and
collects the Authentication options in the order they are defined in the
corresponding Authentication flow.

The new behavior is now easier to understand because the order of the
executions in the authentication flow is now followed exactly when
generating the authentication options.

Note: This might change the order in which authentication options
are presented for some users.

Adapt MultiFactorAuthenticationTest to reflect fixed ordering authenticator ordering (#12102)

The authentication flow for the test
- UsernameFormFactory Required
- SubFlow Required
-- Password Form Alternative
-- OTP Form Alternative

The order of credentials for the "user-with-two-configured-otp" user is:
- otp (first)
- password
- otp (second)

Hence, the test actually tested the broken behavior by expecting the OTPForm after the UsernameForm, because the
authentication options were based on the order of the stored credentials(!) and not in the order
of the authenticators from the authentication flow.

The "correct" order is UsernameForm -> PasswordForm (try another way -> OTP Form), iff the user has a password configured.
The adapted test reflects this new behaviour.

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
@mabartos I just rebased this PR to current master.

Note that the current implementation of [AuthenticationSelectionResolver](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/AuthenticationSelectionResolver.java#L86) uses a deprecated method `SubjectCredentialManager#getConfiguredUserStorageCredentialTypesStream()` that seems to be scheduled for removal. I ignored that for now and also call this method in this PR.

mabartos commented:
@thomasdarimont Could you please check the present test failures in the GH actions?

achojoao commented:
Is this going to be merged any time soon? @thomasdarimont @mabartos  

It will be awesome if we can use it on our project. Thanks!

julienfollow commented:
Did you have a date about the merge of this PR ? 
It's a very useful feature when we have multi authenticator ...

julienfollow commented:
up !

stianst commented:
Shouldn't the order of options presented to the user be based on the user preferences rather than based on realm admins preferences?

achojoao commented:
Well, it depends on the point of view or the specific use case. In my project, I want to be the one to decide the first option to appear. Maybe it could be a configuration parameter or something like that, whether the decision to be based on the user preference or the realm admin, so we can support both approaches.

grishio commented:
> Shouldn't the order of options presented to the user be based on the user preferences rather than based on realm admins preferences?

True in a way. In most cases, if a a global flow was created for a specific pupose, you would expect that flow to be followed as the administrator intends. If your organization is going towards the password-less route, you would want that to be first with password + otp as a backup.

EspenRoth commented:
Have there been any more discussions on this? Any reason to close it? Could even be an optional feature. This would be really nice to have for my team

alexlamazing commented:
@julienfollow can you share more about your plugin?
In my case, I need to show a authentication selection page to users after entering username. They can continue to authenticate according to their selection. Thank you.

stianst commented:
Closing this since it is old and there is no linked issue. If there is still interested please create an issue, and send another PR.
